### PR TITLE
Slimmer docker for ai_inferencer

### PR DESCRIPTION
**Pull request by rzaete**

### CODE REVIEW

**Comment by CESARDELATORRE** [comment]
@rzaete If I compare the requirements.txt from your PR to the requirements.txt from the original model we trained yesterday, it has differences in package versions. Why is that?

Like this:

image

Do we have to use the requirements.txt in your PR?
I think it'd be safer to use a requirements.txt related to the AutoML model trained, each time, because eventually, a new model trained with AutoML could have new/different dependencies, right? - Or at least we need a repeatable procedure...

The  requirements.txt should not be something fixed in the project, I think.

That's why I'm surprised it is part of your PR and question if it should be.
So, what are the reasons for those differences in package versions? 

I see that if I try to use the original requirements.txt from the MLFlow model (instead of your requirements.txt from your PR), I get errors when trying to build the image:

```
84.43       creating build/temp.linux-x86_64-cpython-38/pycocotools
84.43       gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/numpy/core/include -I./common -I/usr/local/include/python3.8 -c ../common/maskApi.c -o build/temp.linux-x86_64-cpython-38/../common/maskApi.o -Wno-cpp -Wno-unused-function -std=c99
84.43       gcc: error: ../common/maskApi.c: No such file or directory
84.43       gcc: fatal error: no input files
84.43       compilation terminated.
84.43       /tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/Cython/Compiler/Main.py:384: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-install-up5bht_y/pycocotools_35745183d44b407e9c3ead05d88a15ec/pycocotools/_mask.pyx
84.43         tree = Parsing.p_module(s, pxd, full_module_name)
84.43       error: command '/usr/bin/gcc' failed with exit code 1
84.43       [end of output]
84.43
84.43   note: This error originates from a subprocess, and is likely not a problem with pip.
84.43   ERROR: Failed building wheel for pycocotools
84.43   Building wheel for recordclass (setup.py): started
87.79   Building wheel for recordclass (setup.py): finished with status 'done'
87.79   Created wheel for recordclass: filename=recordclass-0.14.3-cp38-cp38-linux_x86_64.whl size=425092 sha256=208be5cc382bb2fc57956f9169887ae122c1a4792e0e555b13987b1eff973b67
87.79   Stored in directory: /root/.cache/pip/wheels/f9/37/88/ca3884e0b0afcf25313eb759b61dc3c3cac58e9ef452efeb72
87.79 Successfully built databricks-cli fire fusepy fvcore iopath pretrainedmodels smart-open wrapt recordclass
87.79 Failed to build pycocotools
87.80 ERROR: Could not build wheels for pycocotools, which is required to install pyproject.toml-based projects
------
failed to solve: process "/bin/sh -c pip3 install -r ./mlflow-model/requirements.txt" did not complete successfully: exit code: 1
```

So, if you had to change manually those package versions in order to solve image building issues, then, what's the procedure we should follow every time we get a new trained model that could have new/different package dependencies?

Basically, the dependencies.txt is related to the model to be deployed, so we need to do something repeatable.
If it was a complex process manually changing the package versions, that's a problem for the future models if they change the dependency versions or new packages because we won't be able to use this "fixed requirements.txt" from your PR...

Can you comment of these questions, please? :)

**Comment by rzaete** [comment]
> @rzaete If I compare the requirements.txt from your PR to the requirements.txt from the original model we trained yesterday, it has differences in package versions. Why is that?
> 
> Like this:
> 
> image
> 
> Do we have to use the requirements.txt in your PR? I think it'd be safer to use a requirements.txt related to the AutoML model trained, each time, because eventually, a new model trained with AutoML could have new/different dependencies, right? - Or at least we need a repeatable procedure...
> 
> The requirements.txt should not be something fixed in the project, I think.
> 
> That's why I'm surprised it is part of your PR and question if it should be. So, what are the reasons for those differences in package versions?
> 
> I see that if I try to use the original requirements.txt from the MLFlow model (instead of your requirements.txt from your PR), I get errors when trying to build the image:
> 
> ```
> 84.43       creating build/temp.linux-x86_64-cpython-38/pycocotools
> 84.43       gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/numpy/core/include -I./common -I/usr/local/include/python3.8 -c ../common/maskApi.c -o build/temp.linux-x86_64-cpython-38/../common/maskApi.o -Wno-cpp -Wno-unused-function -std=c99
> 84.43       gcc: error: ../common/maskApi.c: No such file or directory
> 84.43       gcc: fatal error: no input files
> 84.43       compilation terminated.
> 84.43       /tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/Cython/Compiler/Main.py:384: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-install-up5bht_y/pycocotools_35745183d44b407e9c3ead05d88a15ec/pycocotools/_mask.pyx
> 84.43         tree = Parsing.p_module(s, pxd, full_module_name)
> 84.43       error: command '/usr/bin/gcc' failed with exit code 1
> 84.43       [end of output]
> 84.43
> 84.43   note: This error originates from a subprocess, and is likely not a problem with pip.
> 84.43   ERROR: Failed building wheel for pycocotools
> 84.43   Building wheel for recordclass (setup.py): started
> 87.79   Building wheel for recordclass (setup.py): finished with status 'done'
> 87.79   Created wheel for recordclass: filename=recordclass-0.14.3-cp38-cp38-linux_x86_64.whl size=425092 sha256=208be5cc382bb2fc57956f9169887ae122c1a4792e0e555b13987b1eff973b67
> 87.79   Stored in directory: /root/.cache/pip/wheels/f9/37/88/ca3884e0b0afcf25313eb759b61dc3c3cac58e9ef452efeb72
> 87.79 Successfully built databricks-cli fire fusepy fvcore iopath pretrainedmodels smart-open wrapt recordclass
> 87.79 Failed to build pycocotools
> 87.80 ERROR: Could not build wheels for pycocotools, which is required to install pyproject.toml-based projects
> ------
> failed to solve: process "/bin/sh -c pip3 install -r ./mlflow-model/requirements.txt" did not complete successfully: exit code: 1
> ```
> 
> So, if you had to change manually those package versions in order to solve image building issues, then, what's the procedure we should follow every time we get a new trained model that could have new/different package dependencies?
> 
> Basically, the dependencies.txt is related to the model to be deployed, so we need to do something repeatable. If it was a complex process manually changing the package versions, that's a problem for the future models if they change the dependency versions or new packages because we won't be able to use this "fixed requirements.txt" from your PR...
> 
> Can you comment of these questions, please? :)

@CESARDELATORRE Yes, I agree this is not a sustainable solution as adding new dependencies would require a manual operation to merge into the requirements.txt.

The reason I had to include the modified version of the requirements.txt in the PR is the build error in pycocotools 2.0.4 as you stated. I tried different solutions on the web but none of them could fix the error. The package that needs this version of pycocotools is azureml-automl-dnn-vision 1.52.0.post1. After upgrading this package to 1.53.0, the dependency on pycocotools also changed to version 2.0.6 of this package which does not have any build errors and can be installed through pip. Following that I had to change the version of some other packages as well to make them compatible with the new version of azureml-automl-dnn-vision.

For a long-term solution, we have to look into the issue with pycocotools v2.0.4 and potentially try to build it from source and see if that can resolve the issue.

**Comment by CESARDELATORRE** [comment]
> > @rzaete If I compare the requirements.txt from your PR to the requirements.txt from the original model we trained yesterday, it has differences in package versions. Why is that?
> > Like this:
> > image
> > Do we have to use the requirements.txt in your PR? I think it'd be safer to use a requirements.txt related to the AutoML model trained, each time, because eventually, a new model trained with AutoML could have new/different dependencies, right? - Or at least we need a repeatable procedure...
> > The requirements.txt should not be something fixed in the project, I think.
> > That's why I'm surprised it is part of your PR and question if it should be. So, what are the reasons for those differences in package versions?
> > I see that if I try to use the original requirements.txt from the MLFlow model (instead of your requirements.txt from your PR), I get errors when trying to build the image:
> > ```
> > 84.43       creating build/temp.linux-x86_64-cpython-38/pycocotools
> > 84.43       gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/numpy/core/include -I./common -I/usr/local/include/python3.8 -c ../common/maskApi.c -o build/temp.linux-x86_64-cpython-38/../common/maskApi.o -Wno-cpp -Wno-unused-function -std=c99
> > 84.43       gcc: error: ../common/maskApi.c: No such file or directory
> > 84.43       gcc: fatal error: no input files
> > 84.43       compilation terminated.
> > 84.43       /tmp/pip-build-env-8ouhq3qf/overlay/lib/python3.8/site-packages/Cython/Compiler/Main.py:384: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-install-up5bht_y/pycocotools_35745183d44b407e9c3ead05d88a15ec/pycocotools/_mask.pyx
> > 84.43         tree = Parsing.p_module(s, pxd, full_module_name)
> > 84.43       error: command '/usr/bin/gcc' failed with exit code 1
> > 84.43       [end of output]
> > 84.43
> > 84.43   note: This error originates from a subprocess, and is likely not a problem with pip.
> > 84.43   ERROR: Failed building wheel for pycocotools
> > 84.43   Building wheel for recordclass (setup.py): started
> > 87.79   Building wheel for recordclass (setup.py): finished with status 'done'
> > 87.79   Created wheel for recordclass: filename=recordclass-0.14.3-cp38-cp38-linux_x86_64.whl size=425092 sha256=208be5cc382bb2fc57956f9169887ae122c1a4792e0e555b13987b1eff973b67
> > 87.79   Stored in directory: /root/.cache/pip/wheels/f9/37/88/ca3884e0b0afcf25313eb759b61dc3c3cac58e9ef452efeb72
> > 87.79 Successfully built databricks-cli fire fusepy fvcore iopath pretrainedmodels smart-open wrapt recordclass
> > 87.79 Failed to build pycocotools
> > 87.80 ERROR: Could not build wheels for pycocotools, which is required to install pyproject.toml-based projects
> > ------
> > failed to solve: process "/bin/sh -c pip3 install -r ./mlflow-model/requirements.txt" did not complete successfully: exit code: 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, if you had to change manually those package versions in order to solve image building issues, then, what's the procedure we should follow every time we get a new trained model that could have new/different package dependencies?
> > Basically, the dependencies.txt is related to the model to be deployed, so we need to do something repeatable. If it was a complex process manually changing the package versions, that's a problem for the future models if they change the dependency versions or new packages because we won't be able to use this "fixed requirements.txt" from your PR...
> > Can you comment of these questions, please? :)
> 
> @CESARDELATORRE Yes, I agree this is not a sustainable solution as adding new dependencies would require a manual operation to merge into the requirements.txt.
> 
> The reason I had to include the modified version of the requirements.txt in the PR is the build error in pycocotools 2.0.4 as you stated. I tried different solutions on the web but none of them could fix the error. The package that needs this version of pycocotools is azureml-automl-dnn-vision 1.52.0.post1. After upgrading this package to 1.53.0, the dependency on pycocotools also changed to version 2.0.6 of this package which does not have any build errors and can be installed through pip. Following that I had to change the version of some other packages as well to make them compatible with the new version of azureml-automl-dnn-vision.
> 
> For a long-term solution, we have to look into the issue with pycocotools v2.0.4 and potentially try to build it from source and see if that can resolve the issue.

I was guessing that would be your answer.. ;) I agree that for the moment the fixed requerements.txt is the best solution.
We could raise an issue about this to the Azure ML / AutoML computer vision team, too.
In any case, it's clear for me now. :)
